File: BTfromInandPostorder.java
// Space complexity - O(n)
// Space complexity - O(n)
//Did this code successfully run on Leetcode : Yes
//Any problem you faced while coding this : No
class Solution {
public TreeNode buildTreePostIn(int[] inorder, int[] postorder) {
	
	if (inorder == null || postorder == null || inorder.length != postorder.length)
	{
		return null;
		
	}
	HashMap<Integer, Integer> map1 = new HashMap<Integer,Integer>();
	
	for (int i=0;i<inorder.length;++i) 
	{
		map1.put(inorder[i], i);
	}
	return helper(inorder, 0, inorder.length-1, postorder, 0, 
                          postorder.length-1,map1);
}

private TreeNode helper(int[] inorder, int is, int ie, int[] postorder, int ps, int pe, 
                                 HashMap<Integer,Integer> map1){
	
	if (ps>pe || is>ie) {
		return null;
	}
	
	TreeNode root = new TreeNode(postorder[pe]);
	
	int ri = map1.get(postorder[pe]);
	
	TreeNode leftchild = helper(inorder, is, ri-1, postorder, ps, ps+ri-is-1, map1);
	TreeNode rightchild = helper(inorder,ri+1, ie, postorder, ps+ri-is, pe-1, map1);
	
	root.left = leftchild;
	root.right = rightchild;
	
	return root;
}
}

File: BTFromInAndPostOrder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        self.idx = len(inorder) - 1
        self.d = {}
        self.postorder = postorder
        for i,val in enumerate(inorder):
            self.d[val] = i
            
        def helper(postorder,inorder,start,end):
            if start > end:
                return None
            rootVal = postorder[self.idx]
            self.idx -= 1
            root = TreeNode(rootVal)
            pos = self.d[rootVal]
            root.right = helper(postorder,inorder,pos+1,end)
            root.left = helper(postorder,inorder,start,pos-1)
            
            return root
        return helper(postorder,inorder,0,len(inorder)-1)
        